### ISSUE TITLE
ajax_request decorator - use django json encoder

### ISSUE THREAD

**Issue by cezio** [skorokithakis/django-annoying#16]
Current json serializer doesn't support few basic python types. Instead of half-baked serializer used now, maybe Django's encoder class should be used:

https://github.com/django/django/blob/master/django/core/serializers/json.py#L84

This encoder supports additional types:
- datetime,
- date,
- time,
- decimal

**Comment by skorokithakis** [comment]
My concern here is that it will break things for current users. I do agree that the better serializer should be used by default...

**Comment by mknecht** [comment]
How about the usual

**Comment by skorokithakis** [comment]
Hmm, that's an idea, but the new decorator would necessarily be misnamed a bit, so I'm not a huge fan of this practice...

**Comment by mknecht** [comment]
True. How about a configuration option that defaults to the old behavior:

```python
@ajax_request(use_django_encoder=True)
def cost_view(request):
    return {'total': Decimal("4")}

@ajax_request()
def user_view(request):
    return {'id': 7}
```

Arguably, though, it's misnamed already — at least it confused me, since AJAX is just one particular way for a client to access the API. This decorator is really for API views, providing machine-readable representations of a resource, no? I.e. `api_view` or `api_response` might not be a bad fit. Or, more behavior-related names: `encode_response`, `encode_return_value`, … well, just some thoughts. :)

**Comment by skorokithakis** [comment]
The extra parameter still breaks backwards compatibility, since it's not called with parameters by default, but I really like the alternative names. Something like `json_response` might be the most fitting, being very literal.

**Comment by mknecht** [comment]
But it does support YAML, too, no? That's why I shed away from the `JSON` name. :)

Also, Decorators that can be applied with and without arguments are possible, i.e. you don't have to break backwards-compatibility. You “merely” need to consider the type of the first argument, if it's a function or not. If you always use keyword args for the arguments, then it would be perfectly safe, too. Might not be the most beautiful solution on your side of the code, but if you don't want to change the name, and still fix this issue, you can. :)

**Comment by skorokithakis** [comment]
Yeah, you're right. I didn't consider that we'd never be accepting non-function arguments in normal usage, so I'll go with your proposed fix above. If anyone wants to beat me to a PR, you are very welcome to.

**Comment by mknecht** [comment]
Just had to serialize something to JSON, and wanted to try out the built-in serializer. Turns out it is [restricted to Django models](https://docs.djangoproject.com/en/dev/topics/serialization/). Also, a serialize-deserialize does not yield the same objects: For example, decimals stay a string.

So, I don't think this annotation should be switching to Django's serializer. At least I use API views for more than just plain Django models. If I'm missing something, do tell. :)

**Comment by skorokithakis** [comment]
@mknecht, have you found something that Django's JSON encoder won't serialize? Decimals do stay a string, but the default JSON encoder produces an exception, so I'm not sure that's a valid comparison.

**Comment by mknecht** [comment]
@skorokithakis It appears that the serializer will *only* serialize Django models. But, I would like to use the decorator for primitive object graphs, too, like a dict.

The linked documentation indicates that, and I tried and it refused to serialize dicts and such because they did not have a `_meta` attribute.

**Comment by skorokithakis** [comment]
That's odd, I've been using it for proper serialization of dates and it's worked fine. I'll investigate.

**Comment by mknecht** [comment]
Maybe I misunderstood something. Thought it odd, too. I'd be glad to hear it works / what I did wrong. :)

**Comment by mknecht** [comment]
No, you're right, works just fine. I must have messed something up. Sorry for the confusion. :)

**Comment by troyleesmith** [comment]
Yo is it possible to have this issue closed up?

**Comment by skorokithakis** [comment]
Certainly, PRs are welcome. I am a bit swamped right now :/
